test/src/main/java/org/apache/accumulo/test/MaxWalReferencedIT.java

```diff
@@ -0,0 +1,143 @@
+    cfg.setProperty(Property.TSERV_WAL_MAX_SIZE, hdfsMinBlockSize);
+    // Set the max number of WALs that can be referenced
+    cfg.setProperty(Property.TSERV_WAL_MAX_REFERENCED, Integer.toString(WAL_MAX_REFERENCED));
+    cfg.setNumTservers(1);
+
+    // Use raw local file system so WAL syncs and flushes work as expected
+    hadoopCoreSite.set("fs.file.impl", RawLocalFileSystem.class.getName());
+  }
+
+  @Test
+  public void testWALMaxReferenced() throws Exception {
+    try (AccumuloClient client = Accumulo.newClient().from(getClientProperties()).build()) {
+      String tableName = getUniqueNames(1)[0];
+
+      SortedSet<Text> splits = new TreeSet<>();
+      for (int i = 1; i <= 4; i++) {
+        splits.add(new Text(Integer.toString(i)));
+      }
+      client.tableOperations().create(tableName, new NewTableConfiguration().withSplits(splits));
+
+      log.info("Created table {} with splits. Now writing data.", tableName);
+
+      // Write data multiple times until we see the WAL count exceed WAL_MAX_REFERENCED
+      final int rowsPerIteration = 30000;
+      AtomicInteger iteration = new AtomicInteger(0);
+      Wait.waitFor(() -> {
+        int startRow = iteration.get() * rowsPerIteration;
+        int endRow = (iteration.get() + 1) * rowsPerIteration;
+
+        // Write data that should fill or partially fill the WAL
+        writeData(client, tableName, startRow, endRow);
+
+        // Check the current number of WALs in use
+        int walCount = getWalCount(getServerContext());
+        log.info("After iteration {}, wrote rows [{}..{}), WAL count is {}", iteration, startRow,
+            endRow, walCount);
+        iteration.getAndIncrement();
+
+        if (walCount > WAL_MAX_REFERENCED) {
+          log.info("Reached WAL count of {}, now wait for minor compactions to reduce WAL count",
+              walCount);
+          return true;
+        } else {
+          return false;
+        }
+      }, 60000, 250, "Expected to see WAL count exceed " + WAL_MAX_REFERENCED);
+
+      // wait for minor compactions to reduce the WAL count
+      Wait.waitFor(() -> getWalCount(getServerContext()) <= WAL_MAX_REFERENCED, 30000, 1000,
+          "WAL count never dropped within 30 seconds");
+    }
+  }
+
+  private void writeData(AccumuloClient client, String table, int startRow, int endRow)
+      throws Exception {
+    try (BatchWriter bw = client.createBatchWriter(table, new BatchWriterConfig())) {
+      for (int r = startRow; r < endRow; r++) {
+        Mutation m = new Mutation(String.format("row_%07d", r));
+        m.put("cf", "cq", String.format("val_%d", r));
+        bw.addMutation(m);
+      }
+    }
+  }
+
+  private int getWalCount(ServerContext context) throws Exception {
+    WalStateManager wals = new WalStateManager(context);
+    int total = 0;
+    for (Entry<TServerInstance,List<UUID>> entry : wals.getAllMarkers().entrySet()) {
+      total += entry.getValue().size();
+    }
+    return total;
```

Comment:
Write ahead logs can have different states.  For this test do not care about the one in the open state (every tserver will always have 2 open), but want to count the non open ones.

```suggestion
   return wals.getAllState().values().stream().filter(walState -> walState != WalStateManager.WalState.OPEN).count();
```

Reply:
Added in a176b85